pepper-apis/dss-core/src/main/java/org/broadinstitute/ddp/client/Auth0ManagementClient.java

```diff
@@ -527,31 +529,55 @@ private ApiResult<String, APIException> tryPasswordChangeRequest(PasswordChangeT
         });
     }
 
-    private <B, E> ApiResult<B, E> withRetries(String retryMessage, Supplier<ApiResult<B, E>> callback) {
+    public <B, E> ApiResult<B, E> withRetries(String retryMessage, Supplier<ApiResult<B, E>> callback) {
         ApiResult<B, E> res = null;
         int numTries = 0;
         int maxTries = maxRetries + 1;
+        long waitTotal = 0L;
         while (numTries < maxTries) {
             res = callback.get();
             numTries++;
             if (numTries >= maxTries) {
                 break;
             }
-            if (res.getStatusCode() == 429) {
+            long wait = backoffMillis * numTries + new Random().nextInt(MAX_JITTER_MILLIS);
+            if (res.getError() instanceof RateLimitException && res.getStatusCode() == 429) {
                 log.error(retryMessage, res.getError());
-                long wait = backoffMillis * numTries + new Random().nextInt(MAX_JITTER_MILLIS);
+                // Get information from auth0 about how long to wait
+                RateLimitException rateLimit = (RateLimitException) res.getError();
+                long retryAfter = auth0BackoffTime(rateLimit);
+                wait = retryAfter > 0 ? retryAfter : wait;
+                log.warn("Hit Auth0 rate limit: Pausing for " + wait + " milliseconds");
                 try {
                     TimeUnit.MILLISECONDS.sleep(wait);
                 } catch (InterruptedException e) {
                     log.warn("Interrupted while waiting after rate limit", e);
                 }
+                waitTotal += wait;
+                log.warn("Hit Auth0 rate limit: Total wait time is " + waitTotal + " milliseconds");
             } else {
                 break;
             }
         }
         return res;
     }
 
+    /**
+     * A function returns the number of milliseconds to wait for Auth0 rate limit to reset
+     *   or -1 if no time information from Auth0.
+     * @param rateLimitException {@link RateLimitException}
+     */
+    private long auth0BackoffTime(RateLimitException rateLimitException) {
```

Comment:
Now that you have this in a method it would be great to see a unit test to at least verify the positive case.